src/fluree/db/query/fql.cljc

```diff
@@ -1,17 +1,139 @@
+  directly against the index without building full solutions:
+
+    - single triple in :where
+    - single aggregate COUNT(DISTINCT ?s)
+    - no GROUP BY, HAVING, ORDER BY, VALUES, subqueries, or CONSTRUCT
+
+  Note: FROM clauses are ignored when a db is passed directly to query.
+  FROM is only meaningful for query-connection which loads ledgers.
+
+  Returns a spec map {:subject-var \"?s\", :pred-iri <iri>, :alias sym-or-nil}
+  or nil if the optimization does not apply."
+  [q]
+  (when (map? q)
+    (let [{:keys [select where group-by groupBy having order-by orderBy values construct context]} q]
+      (when (and (sequential? select)
+                 (seq select)
+                 ;; Guard against non-seq values in group-by keys (e.g. symbols or errors)
+                 (not (or (and (sequential? group-by) (seq group-by))
+                          (and (sequential? groupBy) (seq groupBy))))
+                 (nil? having)
+                 (nil? (:having q))
+                 (nil? order-by)
+                 (nil? orderBy)
+                 ;; Only treat :values as present if it's a non-empty sequential;
+                 ;; avoid calling seq on non-sequential values.
+                 (not (and (sequential? values) (seq values)))
+                 (nil? construct)
+                 (vector? where)
+                 (= 1 (count where)))
+        (when-let [{:keys [agg-var alias]} (parse-count-distinct-selector select)]
+          (let [subj-var-str (str agg-var)
+                clause       (first where)]
+            (when (and (map? clause)
+                       (= subj-var-str (get clause "@id")))
+              (let [pred-keys (remove #{"@id" "@context"} (keys clause))]
+                (when (= 1 (count pred-keys))
+                  (let [pred-key (first pred-keys)
+                        ;; Expand predicate using the query context so it matches
+                        ;; the IRI stored in flakes. Context must be parsed first.
+                        parsed-ctx (json-ld/parse-context context)
+                        pred-iri   (json-ld/expand-iri pred-key parsed-ctx)]
+                    {:subject-var subj-var-str
+                     :pred-iri    pred-iri
+                     :alias       alias}))))))))))
+
+(defn- count-distinct-subject-fast
+  "Execute a simple COUNT(DISTINCT ?s) query directly against an index without
+  building full query solutions.
+
+  Returns a channel containing a single FQL-style result vector [[n]].
+  Only used when simple-count-distinct-subject-spec has matched."
+  [ds {:keys [pred-iri]}]
+  (go-try
+    (let [;; unwrap dataset if present – for now we only handle a single active
+          ;; graph, which matches the common FROM <graph> use case.
+          db   (if (dataset/dataset? ds)
+                 (or (dataset/get-active-graph ds)
+                     ;; no active graph – behave like empty result
+                     (reduced nil))
```

Comment:
Is it possible to have no active graph? I think it should either be a named graph or default, never a falsey result.